platform/linux-generic/odp_classification.c

```diff
@@ -1555,56 +1555,48 @@ static inline void pmr_debug_print(pmr_t *pmr, cos_t *cos)
 
 /*
  * Match a PMR chain with a Packet and return matching CoS
- * This function gets called recursively to check the chained PMR Term value
- * with the packet.
+ * This function performs a depth-first search in the CoS tree.
  */
-static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, pmr_t *pmr,
-			    odp_packet_hdr_t *hdr)
+static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, odp_packet_hdr_t *hdr)
 {
-	uint32_t i, num_rule;
+	if (!cos || !cos->valid)
+		return cos;
 
-	if (cos == NULL || pmr == NULL)
-		return NULL;
+	while (1) {
+		uint32_t i, num_rule = odp_atomic_load_u32(&cos->num_rule);
 
-	if (!cos->valid)
-		return NULL;
+		for (i = 0; i < num_rule; i++) {
+			pmr_t *pmr = cos->pmr[i];
 
-	if (verify_pmr(pmr, pkt_addr, hdr)) {
-		/* PMR matched */
-		pmr_debug_print(pmr, cos);
+			if (odp_unlikely(!cos->linked_cos[i]->valid))
+				continue;
```

Comment:
This error should never happen, but if it does may be linked cos could be NULL as well. Instead continue, could this be instead: assert(cos->linked_cos[i] != NULL && cos->linked_cos[i]->valid)